Push parquet select to leaves, add correct reordering

nicklan opened the pull request:
This PR makes us finally have correct semantics when we are asked to read a schema that's not the same as the schema in the parquet file.  In particular this adds:

- Correct sorting of columns to match the specified schema (previous code was actually not correctly doing this)
- Identification of leaf fields that are being selected, so we only read exactly what's asked for
- Sorting a struct inside a list.
- Detection of requested columns that don't exist in the parquet file. 
  - If nullable, note and then fill them in with a column of null 
  - Otherwise, error since we're requesting a missing column that can't be null 
- Detection of timestamp columns that need to be cast to match the delta specification, and the code to do the actual casting

This turns out to be _way harder_ than anticipated, so its a lot of code.

Currently does not support reordering things inside `map`s. This is a complex PR as it is, and we don't need support for that just yet. Map content will just be passed through as read from the file.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,900 @@
+                        Arc::new(StructArray::new_empty_fields(0, None));
+                    std::mem::swap(&mut input_cols[parquet_position], &mut placeholder);
+                    // placeholder now holds our struct array that we want to reorder
+                    let struct_array: StructArray = placeholder.into_data().into();
```
note: i'd love to find a cheaper way to go from `Arc<dyn Array>` -> `StructArray`, but haven't found one yet. This isn't to bad (it doesn't copy the data), but really this should just be a cast. Arrow just doesn't have any methods to go to an `Any` (note an `&Any` mind), so we can't downcast it.

nicklan replied:
That does make things cleaner :) I've updated to that

scovich commented on kernel/src/engine/arrow_utils.rs:
If I understand correctly, the problem we face is:
1. The arrow parquet reader _requires_ that the read schema exactly matches the parquet schema. In particular, the read schema cannot have a different field order, nor can it include any extra fields, and any physical fields we want to ignore must be present-but-masked-out.
2. A given parquet file can provide fields in any order, because parquet is not ordinal -- it uses either field names or field ids to identify columns.
3. Thus, we have to convert the read schema from the user into a physical read schema that is compatible with each specific parquet file's physical schema, and _then_ we have to project the resulting data onto the user's read schema (which includes reordering fields and injecting null columns as needed; hopefully we get the masking right and don't have to excise any unwanted columns after the read completes).

If we squint, the above problem is very similar to the one we face with `EngineData` projections -- used for metadata reads -- where the user can ask for a subset of columns, in arbitrary order, and our data visitor has to comply. Our general expression machinery also supports projection -- used for data reads -- so that we can propagate some fields while manipulating or injecting others. 

Thus, I wonder if we should look for ways to beef up and reuse our general projection machinery, rather than adding a third way of doing things? For example, nested column projection support in our expression machinery has been a TODO for a while -- maybe it's time to implement it? 

The one annoyance is that we'd have to repeat this process for every parquet file -- or at least every unique physical schema we encounter -- but this is an arrow-specific situation and arrow doesn't actually have any problem creating expressions at runtime.

<details>

<summary>arrow projection masks</summary>

That said -- The projection support would really only help for the post-read step, for converting the physical data to match the user's read schema. I suspect arrow's [ProjectionMask](https://docs.rs/parquet/latest/parquet/arrow/struct.ProjectionMask.html) will still require some specific and complex code to correctly wire up the pre-read step. 

If I understand correctly, the projection mask is based on the ordinals of an ordered traversal (**) of all fields in the physical schema -- both leaf and internal -- and the mask specifies which of those fields should actually be read. The documentation is underwhelming, but given that it's possible to select roots (even if they are nested columns), I have to assume that any mask that includes the index of an inode will automatically select all leaf fields it contains -- so nested column pruning should _not_ include the parent's index. 

(**) The docs don't say whether the indexing is in-order, pre-order, or post-order! That detail is kind of important! For now I'll assume pre-order, because in-order only makes sense for binary trees (!= schema), and post-order seems a bit unnatural for this case. 

Given the above, I would lean toward a two pass process:
1. Traverse the physical schema to build a namepath -> index map.
2. Traverse the read schema, to convert each read schema namepath to a physical schema index.

For example, given the physical schema:
```
a
  d
  x
b
  y
    z
  e
c
```
Step 1/ above would produce:
```
[a] -> 0,
[a, d] -> 1,
[a, x] -> 2,
[b] -> 3,
[b, y] -> 4,
[b, y, z] -> 5,
[b, e] -> 6,
[c] -> 7
```
And step 2/ would produce the projection mask `[6, 2, 7]` for the read schema below:
```
b
  e
a
  x
c
```
(I'm guessing we actually only need to store leaf indexes, because it's not worth the extra trouble to "optimize" for internal nodes that we fully select?)
</details>

<details>
<summary>engine data projection</summary>

For projections, I could imagine a co-recursion on the read schema and physical data:
* At each internal level, iterate in order over the fields of the read schema
* Look up each field by name in the physical data.
* If the field is present, recurse on it; otherwise, inject a null column in its place
 
</details>

(note: I only skimmed the code very lightly so far, wanting to sort out this question and these ideas before diving into actual code)

(this is really a PR level comment, but dropping it here because github loses top-level comments too easily)

nicklan replied:
This is interesting, but I think unfortunately the use cases are different between `extract/evaluate_expression` and this reordering. I _do_ think we should unify `extract` and `evaluate_expression` though.

The differences come down to:
1. Reading parquet needing to handle potentially missing columns, but `exract/eval` can assume all columns are there (modulo null columns which are a bit tricky to handle)
2. Parquet being actually "flat", so a `struct` column is purely a schema level thing and the actual columns are just flat in the file

So in 
```
a
  d
  x
b
  y
    z
  e
c
```

your above example of selecting:
```
b
  e
a
  x
c
```

the mask of `[2, 6, 7]` is actually _incorrect_ (assuming `a` `b` and `y` are `struct` fields). The correct mask
would be `[1, 3, 4]` because `a`, `b`, and `y` don't contribute to the column indices.


The "trie" idea is actually interesting though. I wrote up a whole big comment where I tried to apply it to all the use cases, in which I realized that there are some fundamental incompatibilities between read/reorder and extract/eval.

I think I could re-write this in terms of building a trie with just the leaf indicies, and then do something like:
```rust
fn get_mask(kernel_schema, trie) -> ProjectionMask {
  let leaves = kernel_schema.leaves(); // get a Vec<Vec<&str>> of all the paths to leaves
  let indexes = leaves.flat_map(|leaf_path| trie.exact_match(leaf_path)); // exact_match returns a Some<usize> if the path was there
  ProjectionMask::leaves(indexes)
}
```

You'd then _like_ to do something like this for column ordering:
```rust
// has_cols being a stand-in for a RecordBatch/StructArray
order_cols(requested_schema, has_cols, trie) -> cols {
  let mut final_cols = vec!()
  for field in requested_schema {
    if let Some(trie_node) = trie.get_child(field.name()) {
      // this was in the parquet, maybe reorder children too
      let parquet_col = has_cols.column(trie_node.value) // THIS IS WRONG
      let col = if trie_node.has_children() {
        StructCol::new(order_cols(field as Schema, parquet_col, trie_node))
      } else {
        parquet_col
      };
      final_cols.push(col);
    } else {
      // not in parquet, push null
      final_cols.push(new_null_col(field.data_type()))
    }
  }
  final_cols
}
```

But as noted in the comment, now that this is dealing with actual hierarchical data, the indicies are different. We could potentially compute a logical and physical index for each Trie node, but it gets a bit messy.

Open to thoughts about how to improve this, and if we want to go with a Trie approach for just the physical mapping part.

One other note: I'd love to upstream this to the parquet reader once we have it working here. I wrote this PR somewhat with that in mind, and think that it could be similar but more efficient if we just had two parquet schemas to reconcile with each other, and with access to some of the schema internals that we would have inside the parquet crate.  But having things be generalized over kernel usage would hinder that.

nicklan replied:
I'll note that what the current code is doing is actually very similar to computing a "trie" for re-ordering, and just skipping the step of flatting and directly returning the mask indicies. So it's likely more efficient that building a "physical+logical" trie would be.

scovich replied:
Will digest further when I have more time, but one question/reaction:
> the mask of [2, 6, 7]`` is actually incorrect (assuming `a` `b` and `y` are struct fields). The correct mask
would be `[1, 3, 4]` because `a`, `b`, and `y` don't contribute to the column indices.

This falls squarely in the "underwhelming documentation" category -- I would also have intuited that only leaf fields should have indexes, because that's how parquet works... but the [ProjectionMask::roots](https://docs.rs/parquet/latest/parquet/arrow/struct.ProjectionMask.html#method.roots) method only makes sense if inodes also have indexes. Otherwise, it would be impossible to express roots for nested schemas. The example in the `ProjectionMask` docs even specifically calls this out (sort of):

<details>

> For example, the schema
> ```
> message schema {
>   REQUIRED boolean         leaf_1;
>   REQUIRED GROUP group {
>     OPTIONAL int32 leaf_2;
>     OPTIONAL int64 leaf_3;
>   }
> }
> ```
> Has roots `["leaf_1", "group"]` and leaves `["leaf_1", "leaf_2", "leaf_3"]`

From that example, I was forced to conclude that if the root set exists (by name, presumably for readability), then there must be corresponding root node indexes. 

Happy to be proven wrong on this tho...
</details>

nicklan replied:
Yeah, I had the exact same thought, and wasted some time in this PR operating under that assumption. So I wrote a program to play around with it and see what happens. Here's the code:
```rust
use std::{fs::File, sync::Arc};

use arrow::{
    datatypes::{DataType, Field, Schema},
    json::ReaderBuilder,
};
use parquet::{
    arrow::{arrow_reader::ParquetRecordBatchReaderBuilder, ArrowWriter, ProjectionMask},
    basic::Compression,
    file::properties::WriterProperties,
};
use serde::Serialize;

#[derive(Serialize)]
struct InnerStruct {
    int32: i32,
    string: String,
}

impl Default for InnerStruct {
    fn default() -> Self {
        InnerStruct {
            int32: 2,
            string: "Default".to_string(),
        }
    }
}

#[derive(Serialize)]
struct OuterStruct {
    inner_a: InnerStruct,
    inner_b: InnerStruct,
    int32: i32,
}

fn do_write() {
    let schema = Schema::new(vec![
        Field::new(
            "inner_a",
            DataType::Struct(
                vec![
                    Field::new("int32", DataType::Int32, false),
                    Field::new("string", DataType::Utf8, false),
                ]
                .into(),
            ),
            false,
        ),
        Field::new(
            "inner_b",
            DataType::Struct(
                vec![
                    Field::new("int32", DataType::Int32, false),
                    Field::new("string", DataType::Utf8, false),
                ]
                .into(),
            ),
            false,
        ),
        Field::new("int32", DataType::Int32, false),
    ]);

    let rows = vec![
        OuterStruct {
            inner_a: InnerStruct {
                int32: 1,
                string: "bar".to_string(),
            },
            inner_b: InnerStruct::default(),
            int32: 1,
        },
        OuterStruct {
            inner_a: InnerStruct {
                int32: 2,
                string: "foo".to_string(),
            },
            inner_b: InnerStruct::default(),
            int32: 2,
        },
    ];

    let mut decoder = ReaderBuilder::new(Arc::new(schema))
        .build_decoder()
        .unwrap();
    decoder.serialize(&rows).unwrap();
    let batch = decoder.flush().unwrap().unwrap();

    let file = File::create("/tmp/foo.parquet").unwrap();

    let props = WriterProperties::builder()
        .set_compression(Compression::SNAPPY)
        .build();

    let mut writer = ArrowWriter::try_new(file, batch.schema(), Some(props)).unwrap();
    writer.write(&batch).expect("Writing batch");
    writer.close().unwrap();
    println!("Wrote foo.parquet");
}

fn do_read(roots: bool, mask: Option<Vec<usize>>) {
    let file = File::open("/tmp/foo.parquet").unwrap();
    let builder = ParquetRecordBatchReaderBuilder::try_new(file).unwrap();
    let builder = if let Some(mask) = mask {
        println!(
            "\nReading with {} mask: {:?}",
            if roots { "roots" } else { "leaves" },
            mask
        );
        let mask = if roots {
            ProjectionMask::roots(builder.parquet_schema(), mask)
        } else {
            ProjectionMask::leaves(builder.parquet_schema(), mask)
        };
        builder.with_projection(mask)
    } else {
        println!("\nReading with no mask");
        builder
    };
    let mut reader = builder.build().unwrap();
    let record_batch = reader.next().unwrap().unwrap();
    arrow::util::pretty::print_batches(&[record_batch]).unwrap();
}

fn main() {
    do_write();
    do_read(false, None);
    do_read(false, Some(vec![1, 4]));
    do_read(true, Some(vec![0, 2]));
}
```

The output is:
```
Wrote foo.parquet

Reading with no mask
+-------------------------+-----------------------------+-------+
| inner_a                 | inner_b                     | int32 |
+-------------------------+-----------------------------+-------+
| {int32: 1, string: bar} | {int32: 2, string: Default} | 1     |
| {int32: 2, string: foo} | {int32: 2, string: Default} | 2     |
+-------------------------+-----------------------------+-------+

Reading with leaves mask: [1, 4]
+---------------+-------+
| inner_a       | int32 |
+---------------+-------+
| {string: bar} | 1     |
| {string: foo} | 2     |
+---------------+-------+

Reading with roots mask: [0, 2]
+-------------------------+-------+
| inner_a                 | int32 |
+-------------------------+-------+
| {int32: 1, string: bar} | 1     |
| {int32: 2, string: foo} | 2     |
+-------------------------+-------+
```

Which makes it pretty clear that `ProjectionMask::roots` are indices for columns at the root level, and will select whole structs etc, while `ProjectionMask::leaves` is the "flat" parquet model. A pretty confusing API for sure.

(if you want to run that to mess around yourself you'll need:
```toml
arrow = { version = "52.0", features = ["prettyprint"] }
parquet = "52.0.0"
serde = { version = "1.0", features = ["derive"] }
```
in your `Cargo.toml`)

scovich replied:
Aside: FYI you can collapse unwieldy sections of code etc by wrapping it in `<details> ... </details>` tags, with an optional `<summary>...</summary>` tag if you want to change the collapsed section's title:

<details>
This is hidden
</details>

<details>
<summary>Hidden section</summary>
Another hidden bit
</details>

(I took the liberty of fixing up your previous comment)

nicklan replied:
> I was suggesting to simply index the data by name, guided by the read schema. 

Yeah, we could do that, but [`column_by_name`](https://docs.rs/arrow-array/52.0.0/src/arrow_array/record_batch.rs.html#324) calls [`Schema::column_with_name`](https://docs.rs/arrow-schema/52.0.0/src/arrow_schema/schema.rs.html#371-374) which uses `find` on its `Fields`, which is `O(n)` for every lookup, and would be expensive for really wide schemas. So I was storing the indices when iterating so we only have to loop over the parquet schema once.

> My suggestions was motivated by the intuition that generating and probing the flat structure is likely a lot smaller/simpler code than what I saw in my first skim of this PR. 

For the mask generation it would be, and then if we want to go with name probing and `O(n^2)` runtime for re-order, it would be all more simple. I think as soon as we want to store indices for reordering, things would be at a similar level of complexity as here.

nicklan replied:
> Aside: FYI you can collapse unwieldy sections of code etc by wrapping it in `<details> ... </details>` tags, with an optional `<summary>...</summary>` tag 

Good to know :)

> (I took the liberty of fixing up your previous comment)

Thanks!

scovich replied:
> `ProjectionMask::roots` are indices for columns at the root level, and will select whole structs etc, while `ProjectionMask::leaves` is the "flat" parquet model. A pretty confusing API for sure.

Oh, you're saying that `roots` and `leaves` are completely different numbering spaces? Where `roots` includes only top-level columns (inner_a: 0, inner_b: 1, int32: 2), while `leaves` includes only leaf columns (int32: 0, string: 1, int32: 2, string: 3, int32: 4)? That's kind of bonkers. Thanks for clarifying.

scovich replied:
> > I was suggesting to simply index the data by name, guided by the read schema. 

> Yeah, we could do that, but [`column_by_name`](https://docs.rs/arrow-array/52.0.0/src/arrow_array/record_batch.rs.html#324) calls [`Schema::column_with_name`](https://docs.rs/arrow-schema/52.0.0/src/arrow_schema/schema.rs.html#371-374) which uses `find` on its `Fields`, which is `O(n)` for every lookup, and would be expensive for really wide schemas. 

🤦 Oh good grief... I forgot that arrow doesn't have O(1) field name lookup like our schema and spark schema do. That definitely makes life harder. Let me chew on that a bit more, but it's likely you're correct that there's no simple(r) way to solve this.

nicklan replied:
> `roots` includes only top-level columns (inner_a: 0, inner_b: 1, int32: 2), while `leaves` includes only leaf columns

Yep, exactly! Bonkers indeed.

> arrow schema doesn't have O(1) field name lookup

Yeah, this is something we could look at fixing upstream, since they could just as well use `IndexMap` like we do and get `O(1)` lookups. But I'd like to get this working in kernel sooner that we could probably merge that upstream.

nicklan replied:
This is basically correct yes. The main thing I'd change is that there _isn't_ such a thing as a "read schema" in the parquet crate. There is _only_ a `ProjectionMask`. So logically yes, we want the schema to read from parquet, but in practice we have to convert to a `Vec` of indices to read.

scovich replied:
We may not ever materialize the read schema, but the arrow data that comes back is structured according to the read schema that can be inferred from the combination of physical schema and projection mask, no? Projection mask is just an ~annoying~indirect way to represent the read schema. And, sadly, it's not stable from file to file because projection masks don't allow column reordering. So that's different from say spark's concept of "read schema."

nicklan replied:
I think this could potentially simplify things. Would you be okay with pushing that off as an issue to look at, so we can get this merged, since we're pretty blocked on a lot of other development by getting this merged?

Once that's done I (or someone) can prototype what a "pull-based" one would look like and see how much simplification we could get.

hntd187 commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,895 @@
+        debug!("Have requested reorder {requested_ordering:#?} on {input_data:?}");
+        let num_rows = input_data.len();
+        let (input_fields, mut input_cols, null_buffer) = input_data.into_parts();
+        let mut final_fields_cols: Vec<FieldArrayOpt> = std::iter::repeat_with(|| None)
```
Is `vec![None; requested_ordering.len()];` not work?

nicklan replied:
Does indeed work. I for some reason didn't do it that way 🤷🏽

hntd187 commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,895 @@
+                ReorderIndex::Child { index, children } => {
+                    let mut placeholder: Arc<dyn ArrowArray> =
+                        Arc::new(StructArray::new_empty_fields(0, None));
+                    std::mem::swap(&mut input_cols[parquet_position], &mut placeholder);
```
This could potentially reorder every field in a given row yes? Would performance be a concern here? it's not unusual to see thousands of columns in a table. Maybe something like the arc swap crate (which says it's more performant, not sure how true that is) to consider here, but I suppose there is no easy in-place style op to do here that would eliminate the placeholder style?

nicklan replied:
Yes it could, but I don't think the performance of swapping two pointers is the issue. What this does is:
- swap two pointers (should be fast)
- call `into_data().into()` on the swapped out pointer (this has some cost, we convert to `ArrayData` and then right back to `StructArray`. I'd like to find a cheaper way to do this, see my comment about wanting a cheaper way to do this as it's actually just a cast)
- re-order the children (this is required, just a recursive call)
- build a new metadata field to match the new order (fairly fast, and has to happen if we've reordered)

Reordering children is just moving around `Arc`s, so it's pretty cheap. So really the only bit that I think could be faster is `into_data().into()` bit.

hntd187 commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,895 @@
+        let fields = Fields::from_iter(
+            final_fields_cols
+                .iter()
+                .map(|fco| fco.as_ref().unwrap().0.clone()), // cheap Arc clone
```
Could we maybe flatten instead of unwrap so we don't potentially panic on failure?

nicklan replied:
Well, if it's `None` something has gone _very wrong_. We _need_ to have grabbed all the input columns and put them in their correct places. Anything left as `None` is a column we've accidentally dropped, so maybe `panic` is the right thing. 

I could change it to an `expect("Accidentally dropped a column when reordering, please report a bug against kernel")` if you think that would be better :)

nicklan replied:
No more panic.  Have added the subtask.

scovich reviewed (commented):
Flushing comments that have accumulated for a couple days now, in hopes that they're useful.

High level -- I _think_ I understand the algo now, but there's no doc comment nor example anywhere that would make it easier to understand and verify the code. Strongly suggest to add something that shows what is being accomplished.

nicklan commented on kernel/src/engine/default/parquet.rs:
```diff
@@ -141,8 +141,9 @@ impl FileOpener for ParquetOpener {
-                rbr.map_err(Error::Parquet)
-                    .and_then(|rb| reorder_record_batch(rb, &requested_ordering))
+                rbr.map_err(Error::Parquet).and_then(|rb| {
+                    reorder_struct_array(rb.into(), &requested_ordering).map(|sa| sa.into())
```
Yep, works, thanks

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,994 @@
+
+    use super::{get_requested_indices, reorder_struct_array, ReorderIndex};
+
+    macro_rules! rii {
+        ($index: expr) => {{
+            ReorderIndex::Index { index: $index }
+        }};
+    }
```
it does indeed. need to remember that's in the rust toolbox now :)

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,994 @@
+        ));
+        let list =
+            GenericListArray::try_new(new_list_field, offset_buffer, result_array, null_buf)?;
+        Ok(Some((new_field, Arc::new(list))))
```
no good reason, just from moving code around. fixed.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,994 @@
+    list_array: GenericListArray<O>,
+    input_field_name: &str,
+    children: &[ReorderIndex],
+) -> DeltaResult<FieldArrayOpt> {
```
It lets me reuse the type alias. I think this looks nicer than `DeltaResult<(Arc<ArrowField>, Arc<dyn ArrowArray>)`, and if this didn't return an option, I'd just have to wrap it in one at the call site since the result is put into `final_fields_cols` which is holding options.

scovich replied:
Gotcha. Tho an argument could be made that the code is easier to read/maintain if the `Some` is at the use site instead of returned from a function, I'm not sure it's enough to be worth the hassle.

nicklan replied:
Agree, and if it was more generally used I'd replace it, but since it's really only ever going to be called here, I think this is okay. Happy to change it if you feel strongly though :)

scovich replied:
No need, I was just brainstorming.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,995 @@
+    }
+}
+
+// count the number of physical columns, including nested ones in an `ArrowField`
+fn count_cols(field: &ArrowField) -> usize {
+    _count_cols(field.data_type())
+}
```
Yep, it is. I wrote this inspired by the [`fields`](https://docs.rs/arrow-schema/52.1.0/src/arrow_schema/field.rs.html#362) method in arrow. If it was publicly exposed I'd certainly want a trait, but since this is the only usage I thought the parsimonious (lovely word btw) solution was better.

scovich replied:
Ack. Especially since the trait wouldn't allow the other simplification below.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,995 @@
-/// Create a mask that will only select the specified indicies from the parquet. Currently we only
+/// Create a mask that will only select the specified indices from the parquet. Currently we only
 /// handle "root" level columns, and hence use `ProjectionMask::roots`, but will support leaf
 /// selection in the future. See issues #86 and #96 as well.
```
Yep, I did actually update the comment, but I guess not before you looked at this :)

For the args, I've kept them in because of the TODO. I'd like to run some tests to see if there's any different between not setting a `ProjectionMask` vs. a `leaves` mask that has all the fields.

If it's different, I'd add an optimization here to detect that and return `None`, if there's no real difference, I'll remove the args and the optional part of the return type. I just figured I wouldn't bother to make all the changes to the call sites until I was sure which way to go.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,995 @@
+                    }
+                }
+            }
+            _ => {
+                if let Some((index, _, requested_field)) =
+                    requested_schema.fields.get_full(field.name())
+                {
```
🤦🏽, you did not miss it. it's currently in the expression eval only, and I neglected to bring it over. will add

nicklan replied:
added

scovich commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,995 @@
+
+fn _count_cols(dt: &ArrowDataType) -> usize {
+    match dt {
+        ArrowDataType::Struct(fields) => fields.iter().fold(0, |acc, f| acc + count_cols(f)),
+        ArrowDataType::Union(fields, _) => fields.iter().fold(0, |acc, (_, f)| acc + count_cols(f)),
```
Gotcha, thanks for clarifying.

sppalkia reviewed (commented):
Flushing out some minor comments and a question. I read through the larger thread between yourself and Ryan,  my main confusion was around the mask indices and what they're used for -- I think it's confusing me because it looks like the Parquet schema isn't using the Thrift-style representation but rather some intermediate thing instead? This part especially confused me:

> the mask of [2, 6, 7]`` is actually incorrect (assuming a `b` and `y` are struct fields). The correct mask
would be `[1, 3, 4]` because `a`, `b`, and `y` don't contribute to the column indices.

Because in Parquet the nested schema elements do occupy an "index" in the schema list.

hntd187 commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,994 @@
-        Error::generic("Didn't find all requested columns in parquet schema")
-    );
-    Ok((mask_indicies, reorder_indicies))
+) -> DeltaResult<(Vec<usize>, Vec<ReorderIndex>)> {
```
Hmmm, I don't know if rolling this isn't a struct for internal use would make sense. Especially since you are likely going to just assign it in something like `let (mask, reordered_indexes) = get_requested_indices(...);`

sppalkia replied:
Makes sense, though you can always de-structure structs too:

```rust
let SchemaIndex { mask, reordered_indices } = get_requested_indices(...);
```

nicklan replied:
Yeah, given this purely internal for now, I think leaving it as a tuple with a clear doc comment is okay, but lmk if you feel strongly about it :)

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,995 @@
+                    requested_schema.fields.get_full(field.name())
+                {
+                    match (key_val_field.data_type(), requested_field.data_type()) {
+                        (ArrowDataType::Struct(inner_fields), DataType::Map(map_type)) => {
```
yep. the arrow crate [states](https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.html#variant.Map):

> In a field with Map type, the field has a child Struct field, which then has two children: key type and the second the value type. The names of the child fields may be respectively “entries”, “key”, and “value”, but this is not enforced.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,995 @@
+    /// reorder_struct_array to insert the null col
+    fn is_ordered(&self) -> bool {
+        match self {
+            ReorderIndex::Child { ref children, .. } => is_ordered(children),
```
indeed it is not

nicklan commented:
> my main confusion was around the mask indices and what they're used for -- I think it's confusing me because it looks like the Parquet schema isn't using the Thrift-style representation but rather some intermediate thing instead? This part especially confused me:
> 
> > the mask of [2, 6, 7]`` is actually incorrect (assuming a `b` and `y` are struct fields). The correct mask
> > would be `[1, 3, 4]` because `a`, `b`, and `y` don't contribute to the column indices.
> 
> Because in Parquet the nested schema elements do occupy an "index" in the schema list.

Yeah, so the parquet crate is just using a "flat" representation. The method we call is [`ProjectionMask::leaves`](https://docs.rs/parquet/latest/parquet/arrow/struct.ProjectionMask.html#method.leaves). The docs are woefully incomplete for that method, but basically anything that's logical (i.e. nested) "doesn't count". So something like:
```
struct { colA, colB }, struct { colC, colD }
```

Is actually only _4_ columns in the parquet (assuming those cols aren't nested themselves). But the schema for this would be something with 2 columns at the root and then each child having 2 columns. So the structs themselves "don't count" when trying to build a list of indices that we want to select out of the file. 

If we wanted only `colA` and `colC` we'd need to pass `[0, 2]` to the `ProjectionMask::leaves` call, meaning we "skip" over the entries in the schema describing the Struct, since it's purely logical. Hopefully that makes some sense?

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,994 @@
+                        field: Arc::new(field.try_into()?),
+                    });
+                } else {
+                    return Err(Error::Generic(format!(
```
I did rename `Null` to missing. But I think in this context we do want to return an error, because you're asking us to read something that's not there. If it's something with a default value, it shouldn't be passed in the requested schema.

Perhaps at some point we could support passing default values in here, and then this code could fill it in, but forcing callers to deal with `Missing` at any arbitrary depth in a complex parquet record batch seems... not so nice.

I'd rather just only have this code support adding in the null columns for nullable stuff, and punt defaults to the future, if they are needed.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,994 @@
+                    parquet_offset += count_cols(field) - 1;
+                }
+            }
+            ArrowDataType::List(list_field)
+            | ArrowDataType::LargeList(list_field)
+            | ArrowDataType::ListView(list_field) => {
+                if let Some((index, _, requested_field)) =
```
Luckily (I think), we don't need to deal with too much about how stuff is actually represented here. We just want to make sure there is _logical_ agreement between the schemas.

We don't have to convert because we get an [`ArrowReaderMetadata`](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/struct.ArrowReaderMetadata.html) from the file and then call [`schema`](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/struct.ArrowReaderMetadata.html#method.schema) on it, which gives us back an already converted Arrow schema.

nicklan replied:
Discussed offline. Our new casting infra can deal with these cases as they come up. Currently don't anticipate this being needed for lists.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,1180 @@
+    match (kernel_type, arrow_type) {
+        (DataType::Primitive(PrimitiveType::Timestamp), ArrowDataType::Timestamp(_, _))
+        | (DataType::Primitive(PrimitiveType::TimestampNtz), ArrowDataType::Timestamp(_, _)) => {
+            // We assume that any timestamp data read from a delta table is correctly written in
```
note: this is... questionable. we should not do this long term, but i need to figure out what's going on on mac such that things read incorrectly, and then if it really is reading in nanoseconds for some reason, we'd need to do some kind of conversion on the fly (maybe).

Regardless, not something I want to tackle in this PR.

Note that the file in question is `acceptance/tests/dat/out/reader_tests/generated/all_primitive_types/delta/part-00000-a31beaf2-ba0b-4ad6-a402-e5867ba52c91-c000.snappy.parquet` and it really does seem to have things stored in microseconds with UTC

<details>
  <summary>Running `parquet-tools summary` on the file</summary>

```
$ parquet-tools inspect part-00000-a31beaf2-ba0b-4ad6-a402-e5867ba52c91-c000.snappy.parquet
############ file meta data ############
created_by: parquet-mr version 1.13.1 (build db4183109d5b734ec5930d870cdae161e408ddba)
num_columns: 12
num_rows: 5
num_row_groups: 1
format_version: 1.0
serialized_size: 2189


############ Columns ############
utf8
int64
int32
int16
int8
float32
float64
bool
binary
decimal
date32
timestamp

############ Column(utf8) ############
name: utf8
path: utf8
max_definition_level: 1
max_repetition_level: 0
physical_type: BYTE_ARRAY
logical_type: String
converted_type (legacy): UTF8
compression: SNAPPY (space_saved: -2%)

...[snip]...

############ Column(timestamp) ############
name: timestamp
path: timestamp
max_definition_level: 1
max_repetition_level: 0
physical_type: INT64
logical_type: Timestamp(isAdjustedToUTC=true, timeUnit=microseconds, is_from_converted_type=false, force_set_converted_type=false)
converted_type (legacy): TIMESTAMP_MICROS
compression: SNAPPY (space_saved: 0%)
```

</details>

But the mac tests fail if I don't use this hack: https://github.com/delta-incubator/delta-kernel-rs/actions/runs/9882874132/job/27296554771 complaining  `"Incorrect datatype. Expected Timestamp(Microsecond, Some(\"UTC\")), got Timestamp(Nanosecond, None)"`

nicklan replied:
Ahh, nm, this is an issue with DAT. My local version works because the file there has the correct format, but the one the test runners pull is incorrect. I will figure out how to fix that and revert this hack.

nicklan replied:
Okay, this turned out to be something that's fixed in the newest version of `dat`, so I've reverted the hack, and bumped the version for our acceptance tests.

Created #279 to follow up on this

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,994 @@
+            })?;
+            field_vec.push(field);
+            reordered_columns.push(array);
+        }
```
Yep, thanks. Had to annotate the type on `field_vec` since we call `len()` on it before we use it in a context where the type is known.

nicklan commented:
> High level -- I _think_ I understand the algo now, but there's no doc comment nor example anywhere that would make it easier to understand and verify the code. Strongly suggest to add something that shows what is being accomplished.

I've added a big doc comment that tries to cover how things work. LMK if that helps.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,994 @@
+        assert_eq!(ordered_s1.column_names(), vec!["c", "b"]);
+    }
+
+    #[test]
```
Yep, added

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,995 @@
+        | ArrowDataType::LargeList(field)
+        | ArrowDataType::FixedSizeList(field, _)
+        | ArrowDataType::Map(field, _) => count_cols(field),
+        ArrowDataType::Dictionary(_, value_field) => _count_cols(value_field.as_ref()),
```
I'm not planning to use the trait thing for now, but lmk if you'd really like that

scovich replied:
Totally your call, I was just brainstorming

nicklan commented:
@scovich @sppalkia thanks for the reviews! I think I've addressed all the comments, so re-requesting. Apologies if I missed anything, github makes it hard to be sure.

Please do have a look at [this](https://github.com/delta-incubator/delta-kernel-rs/pull/271/files#r1673216002) new thing. Moving the type verification in uncovered something strange in parquet reading on MacOS. I've papered over it for now, but definitely need to look into it further. Will add an issue here if we're all okay with moving forward with this somewhat hacky solution for now.

scovich commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,995 @@
+
+    /// check if this indexing is ordered. an `Index` variant is ordered by definition. a `Null`
+    /// variant is not because if we have a `Null` variant we need to do work in
+    /// reorder_struct_array to insert the null col
```
Ack. That makes more sense, thanks.

nicklan commented:
> Moving the type verification in uncovered something strange in parquet reading

It's that spark sometimes writes `INT96` timestamps (which are [deprecated](https://issues.apache.org/jira/browse/PARQUET-323)). The parquet crate defaults these to Nanosecond resolution, so we get a schema mismatch.

I've partially fixed that via updating the DAT version, but what's funny is that now a test I added just today fails, for the same reason. This is with data pulled _directly_ from a delta-sharing table, so I'm worried we are going to encounter this in the wild often enough to need to do something about it.

I'm considering trying to figure out if the data type is `INT96` on a column, and then truncate the timestamp to micros somehow, but I'm not totally clear how easy that will be.  That is what the Java folks do, so we may need to do the same.

nicklan commented:
Quick follow-up for anyone about to review:

After discussing with Shoumik, the plan is to detect if two data-types are compatible (i.e. the INT96 <-> Microsecond) case, and if so, to allow the read to proceed, and then to cast to the expected type before finally returning from the read.

Initially planning to support only the timestamp one, and decimals, but can add more as needed.

nicklan commented on acceptance/src/data.rs:
```diff
@@ -60,11 +60,12 @@ pub fn sort_record_batch(batch: RecordBatch) -> DeltaResult<RecordBatch> {
+    // INT96 (which is nanosecond precision), while the spec says we should read partition columns
+    // as microseconds. This means the read and golden data don't line up. When this is released in
+    // `dat` upstream, we can stop skipping these tests
+    "all_primitive_types",
```
Now that we handle timestamps correctly, the golden data is "wrong" for this table. This will be fixed by updating to data 0.0.3, and I'll do that shortly after this merges. I've verified manually that we correctly read these two tables.

nicklan commented on kernel/tests/read.rs:
```diff
@@ -511,8 +513,7 @@ fn read_with_scan_data(
     if expected.is_empty() {
         assert_eq!(batches.len(), 0);
     } else {
-        let schema = batches[0].schema();
```
small bug fix here. arrow will assume a column isn't nullable if it doesn't find any nulls in it. so using the schema of the first batch as the schema for all is incorrect if the first batch happens to have a column with no nulls, but subsequent batches DO have nulls. Instead we do the correct thing now and transform the schema the scan promises to read into an arrow schema, and use _that_ as the schema for the batches.

nicklan commented on kernel/tests/read.rs:
```diff
@@ -987,18 +988,60 @@ fn with_predicate_and_removes() -> Result<(), Box<dyn std::error::Error>> {
     read_table_data_str("./tests/data/with-short-dv/", None, None, expected)?;
     Ok(())
 }
+
```
these are taken from the delta-java tests. i validated that the `expected` arrays here are correct.

scovich commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,1338 @@
+            // build a list of kernel fields that matches the order of the arrow fields
+            let mapped_fields = arrow_fields
+                .iter()
+                .flat_map(|f| kernel_fields.fields.get(f.name()));
```
nit: [filter_map](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map) is probably clearer than `flat_map` here? We're specifically trying keep the content of `Some` entries corresponding to successful name lookups, right?

scovich commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,1338 @@
+                            fields,
+                            mask_indices,
+                        )?;
+                        // advance the number of parquet fields, but subtract 1 because the
+                        // struct will be counted by the `enumerate` call but doesn't count as
+                        // an actual index.
+                        parquet_offset += parquet_advance - 1;
```
Just to verify: This also works for deeper schemas because each level of nesting introduces a new `Struct` which will also decrement the count appropriately?

nicklan replied:
yep, exactly

scovich commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,1338 @@
+    /// For a non-nested type, indicates that we need to cast to the contained type
+    Cast(ArrowDataType),
+    /// Used for struct/list/map. Potentially transform child fields using contained reordering
+    Child(Vec<ReorderIndex>),
```
Wording seems a bit off here -- seems like "nested" or "children" or even "parent" would be more accurate than "child" ?

scovich commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,1338 @@
+
+impl ReorderIndex {
+    fn new_cast(index: usize, target: ArrowDataType) -> Self {
+        ReorderIndex {
+            index,
+            transform: ReorderIndexTransform::Cast(target),
+        }
```
nit: can shave line count a bit by pulling out a value:
```suggestion
        let transform = ReorderIndexTransform::Cast(target);
        ReorderIndex { index, transform }
```
or even add and use a (possibly private) `new` method:
```suggestion
        ReorderIndex::new(index, ReorderIndexTransform::Cast(target))
```

(more below)

scovich commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,1338 @@
+                        return Err(Error::unexpected_column_type(field.name()));
+                    }
+                } else {
+                    // We're NOT selecting this field, but we still need to update how much we skip
```
```suggestion
                    // We're NOT selecting this field, but we still need to track how many 
                    // leaf columns we skipped over
```

scovich commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,1338 @@
+            "Getting indices for field {} with offset {parquet_offset}, with index {parquet_index}",
+            field.name()
+        );
+        match field.data_type() {
```
aside: This is a really large match... 200+ LoC if I count correctly. I wish there were better ways to split up such code in rust, but that's a bit trickier given that enum variants are not types and so can't be part of a helper method signature... I suppose helper methods could take the deconstructed enum variant content instead, but then you lose some semantic info.

scovich commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -1,89 +1,1338 @@
+            "Getting indices for field {} with offset {parquet_offset}, with index {parquet_index}",
+            field.name()
+        );
+        match field.data_type() {
+            ArrowDataType::Struct(fields) => {
+                if let Some((index, _, requested_field)) =
+                    requested_schema.fields.get_full(field.name())
+                {
```
If I'm not mistaken, we currently have significant code replication, because every match arm includes the same requested field extraction:
```rust
                if let Some((index, _, requested_field)) =
                    requested_schema.fields.get_full(field.name())
                {
                    ...
                }
```
seems better to do the extraction first, and then match only if successful?
```rust
        if let Some((index, _, requested_field)) =
            requested_schema.fields.get_full(field.name())
        {
            match field.data_type() {
                ArrowDataType::Struct(fields) => { ... }
                  ...
                _ => { ... }
            }
        }
```
could even take it a step further with a `filter_map` call:
```rust
    let requested_fields = fields
        .iter()
        .enumerate()
        .filter_map(|parquet_index, field| {
            debug!(
                "Getting indices for field {} with offset {parquet_offset}, with index {parquet_index}",
                field.name(),
            )
            requested_schema.fields.get_full(field.name())
        });
    for (index, _, requested_field) in requested_fields {
```

nicklan replied:
Yeah, I was able to improve this thanks to this suggestion. Couldn't quiet do what you suggested because even for unselected fields we need info (so we can skip their leaf columns), but now I can make an iterator that gets out the needed info, and then iterate over that and reduce the match complexity quite a bit.

scovich reviewed (commented):
Flushing a bunch of comments. They almost all relate to readability and maintainability rather than correctness -- this code is so complex, big, and subtle that readability is paramount.

Unfortunately I'm not seeing any opportunities for significant simplifications, so we're reduced to this...

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -228,9 +455,18 @@ fn get_indices(
+                        DataTypeCompat::NeedsCast(target) =>
+                            reorder_indices.push(ReorderIndex::new_cast(index, target)),
+                        DataTypeCompat::Nested => return
+                            Err(Error::generic(
```
Good idea. Added an "InternalError" error type. And yeah, anything that would be `Nested` should have been handled above. We only have Nested for the expression evaluator that doesn't care about the return type and just that it's not an `Err`. (Although at some point that might care about casting, so we can extend it then, not needed in this PR).

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -145,39 +381,34 @@ fn get_indices(
+                            ));
                         }
+                        // safety, checked that we have 1 element
+                        let mut children = children.into_iter().next().unwrap();
```
Yep, that works thanks. We do have to make it `mut`. Even though we own it, we still can't modify things that aren't mut.

nicklan commented on kernel/src/engine/arrow_utils.rs:
```diff
@@ -15,54 +16,292 @@ use arrow_schema::{
+* 1. First check if we're already in order (see doc comment for `is_ordered`)
+* 2. If ordered we're done, return, otherwise:
+* 3. Create a Vec[None, ..., None] of placeholders that will hold the correctly ordered columns
+* 4. Deconstruct the existing struct array and then loop over the `ReorderIndex` list
```
Yep, docs updated

scovich reviewed (approved):
LGTM!
